### PR TITLE
core/bloombits, eth/filters: handle null topics

### PR DESCRIPTION
**Pull request by karalabe**
When implementing the new bloombits based filter, I've accidentally broke `null` topics by removing the special casing of `common.Hash{}` filter rules (https://github.com/ethereum/go-ethereum/pull/14631/commits/f585f9eee8cb18423c23fe8b517b5b4cbe3b3755#diff-398162818e80393efa0b6f121150278bL205), which acted as the wildcard topic until now.

This PR fixes the regression, but instead of using the magic hash `common.Hash{}` as the `null` wildcard, the PR reworks the code to handle `nil` topics during parsing, converting a JSON `null` into nil `[]common.Hash` topic.

### CODE REVIEW

File: core/bloombits/matcher.go [hunk 1]
```diff
@@ -95,11 +96,19 @@ func NewMatcher(sectionSize uint64, filters [][][]byte) *Matcher {
...
 		for i, clause := range filter {
+			if clause == nil {
+				bloomBits = nil
+				break
+			}
 			bloomBits[i] = calcBloomIndexes(clause)
 		}
-		m.filters = append(m.filters, bloomBits)
+		// Accumulate the filter rules if no nil rule was within
+		if bloomBits != nil {
```
Comment (zsfelfoldi) [inline core/bloombits/matcher.go]: You could drop this check if you continue-d the outer loop instead of just break-ing out of the inner loop.

Comment (karalabe) [reply core/bloombits/matcher.go]: I don't really like it when multiple nestings are broken out of. Seems a lot harder to follow what's going on, especially since the label can be arbitrarily far away.

Comment (Arachnid) [reply core/bloombits/matcher.go]: I kind of agree with both positions, but in this case personally I think it'd be clearer if it did continue the outer loop; the code is compact, so it's clear what the label is referring to. Alternately, you could refactor the inner loop and append into a separate function.

File: interfaces.go [hunk 2]
```diff
@@ -146,7 +146,7 @@ type FilterQuery struct {
 	// {{}, {B}}          matches any topic in first position, B in second position
 	// {{A}}, {B}}        matches topic A in first position, B in second position
 	// {{A, B}}, {C, D}}  matches topic (A OR B) in first position, (C OR D) in second position
-	Topics [][]common.Hash
+	Topics [][]*common.Hash
```
Comment (fjl) [inline interfaces.go]: I don't get why this change is needed. The wildcard for a topic is the nil slice as documented just above this line. `(anything OR XXX)` is the same as `anything`.

Comment (karalabe) [reply interfaces.go]: [][]common.Hash means

[t1 OR t2] AND [t3 or t4]

You can't set t1 or t2 or t3 or t4 to nil, only the entire group.

Comment (karalabe) [reply interfaces.go]: Ah, you mean interpreting the `null`s already during parsing and getting rid of them.

**Comment by karalabe** [comment]
Simplified according to @fjl's request. @Arachnid @fjl @zsfelfoldi PTAL
